### PR TITLE
singularity-tools: make runscript modifiable

### PR DESCRIPTION
**Pull request by ShamrockLee**
## Description of changes

This pull request makes `.singularity.d/runscript` in the image sandbox modifiable by copying the content instead of linking, a quick fix to the build failure encountered when updating singularity to 4.1.5 (#334789).

Currently, script writers in `writers` don't expose the `.text` attribute, as trivial helpers derived from `writeTextFile` do. Once we have it, we could obtain references to the content of `runScriptFile` without actually building it using `lib.splitString "\n" (writeStringReferencesToFile writeScriptFile.text).text`.

Cc: @GaetanLepage @SomeoneSerge

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ShamrockLee** [comment]
@ofborg build apptainer.tests.image-hello-cowsay
@ofborg build singularity.tests.image-hello-cowsay

**Review by GaetanLepage — APPROVED** [review]
LGTM ! Thanks !

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 334826` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 334826` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/build-support/singularity-tools/default.nix [hunk 1]
```diff
@@ -82,7 +83,17 @@ rec {
...
-            layerClosure = writeClosure contents;
+            layerClosure = writeClosure (
+              [
+                bashInteractive
+                # runScript references
+                # TODO(@ShamrockLee):
+                # Implement fine-grain reference parsing with writeStringReferencesToFile
+                # once we make writers.makeScriptWriter expose the .text attribute.
+                (writeDirectReferencesToFile runScriptFile)
+              ]
```
Comment (SomeoneSerge) [inline pkgs/build-support/singularity-tools/default.nix:94]: I didn't understand the comment or how this bit relates to the PR description. Why two layers of derivations for extracting the references?

**Comment by SomeoneSerge** [comment]
Making the runscript writeable doesn't sound unreasonable per se (e.g. I imagine we could one day with to have a hook to extend it with some information that's only available at build time, not eval time), but is this the right approach to address the original issue?

```
INFO:    Inserting Singularity configuration...
FATAL:   While performing build: packer failed to pack: while inserting base environment: build: failed to make environment files: chmod /tmp/disk/build-temp-577913542/rootfs/.singularity.d/runscript: operation not permitted
[    4.385199] reboot: Power down
```


What does it mean that singularity is trying to "insert base environment", and why?

File: pkgs/build-support/singularity-tools/default.nix (hunk 1 above)
Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix:94]: Here's what I would do instead if we have `runScriptFile.text`:

```nix
runCommand {
  layerClosure = writeClosure ([
    bashInteractive
  ] ++ lib.splitString "\n" (writeStringReferencesToFile runScriptFile.text).text)
  runScriptText = runScriptFile.text;
  passAsFile = [ "runScriptFile" ]
} ''
  ...
  mv "$runScriptText" singularity.d/runscript
''
```

This way, we don't need to instantiate and build the intermediate derivation `runScriptFile` while still getting it into the container. We don't even have to write the `runScript` content twice.

Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix:94]: We could achieve this by switching back to `writeShellScript` to construct `runScriptFile`, but I prefer to fix the `writer.makeScriptWriter` to keep using the unified interface of `writers.writeBash`.

**Comment by ShamrockLee** [comment]
I still need to check that out.

Still,

```
chmod /tmp/disk/build-temp-577913542/rootfs/.singularity.d/runscript: operation not permitted
```

Suggests that the error happens when SingularityCE tries to `chmod` the `.singularity.d/runscript` and experiences permission denial. This is because it is a symbolic link to the container Nix store in the current implementation.

This PR makes it a regular file instead of a symbolic link. If that is not enough, we could add a `chmod u+w` line.

**Comment by ShamrockLee** [comment]
@GaetanLepage Does the updated singularity build the test image after applying this PR?

**Comment by GaetanLepage** [comment]
> @GaetanLepage Does the updated singularity build the test image after applying this PR?

Indeed it does !

```
/nix/store/y68xf4gccqjr0giqv0nlm6j1d02asbab-singularity-image-hello-cowsay.sif
```

File: pkgs/build-support/singularity-tools/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/build-support/singularity-tools/default.nix]: Hmm what's `(writeStringReferencesToFile runScriptFile.text).text`

**Comment by SomeoneSerge** [comment]
> Indeed it does !

And what's the diff between the original runscript and the mutated runscript?

File: pkgs/build-support/singularity-tools/default.nix (hunk 1 above)
Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix]: > Hmm what's `(writeStringReferencesToFile ...).text`

If `builtins.getContext` is available, writeStringReferencesToFile will use it to get the reference paths, compose to a string, append the string context, and write the string with `writeText`. `writeText` exposes the input string through the `text` attribute.

Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix]: It turns out to be hard to make `writers.makeScriptWriter` expose the `text` attribute, as it works around Darwin's prohibition of script-as-shebang-interpreter by determining at build time if the input `interpreter` is a script and constructing a custom shebang.

I'll switch it back to `writeScript` then.

Comment (SomeoneSerge) [reply pkgs/build-support/singularity-tools/default.nix]: > If builtins.getContext is available, writeStringReferencesToFile will use it to get the reference paths,

If I understood you correctly, this means "all store paths from `inputDrvs` (=direct dependencies, not transitive) referenced by the environment or the builder script (even those paths that are not referenced in the output)"? If so, why that? I'm missing the high level picture

> string context

Oy vey, I'd rather not rely on string contexts. E.g. TVix will not implement them

Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix]: It's `runInLinuxVM`'s fault again. It doesn't respect the `passAsFile` specification of the input attribute (#16742). Building and copying `runScriptFile` seems to be the most sensible implementation.

Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix]: I just pushed the revised version. Please take a look.

File: pkgs/build-support/singularity-tools/default.nix [hunk 2]
```diff
@@ -85,8 +92,8 @@ lib.makeExtensible (final: {
             layerClosure = writeClosure (
               [
                 bashInteractive
-                runScriptFile
               ]
+              ++ runScriptReferences
```
Comment (SomeoneSerge) [inline pkgs/build-support/singularity-tools/default.nix:96]: I think I finally get it: your aim is to get `runScriptFile`'s dependencies into the image, without holding two copies of the runscript (one in `.singularity.d` and one in `/nix/store`)?

File: pkgs/build-support/singularity-tools/default.nix (hunk 1 above)
Comment (ShamrockLee) [reply pkgs/build-support/singularity-tools/default.nix]: > > string context
>
> Oy vey, I'd rather not rely on string contexts. E.g. TVix will not implement them

Thanks for letting me know. The new implementation checks whether `builtins.getContext` is supported before choosing the referencing approach.

**Comment by GaetanLepage** [comment]
Don't forget to format your files with `nixfmt-rfc-style`.

**Comment by GaetanLepage** [comment]
Should this be merged now ? Any remaining blocker ?

**Comment by ShamrockLee** [comment]
> And what's the diff between the original runscript and the mutated transcript?

@SomeoneSerge It's an interesting question, but before #268199, there was no easy way to override the container build process.

Here's the result obtained by cherry-picking #334789 and adding the following lines at the bottom of `singularity-tools.bulidImage` container build commands:
```
            stat img/.singularity.d/runscript
            diff -u ${runScriptFile} img/.singularity.d/runscript || true
```

The stat result is as follows:

```
  File: img/.singularity.d/runscript
  Size: 97              Blocks: 8          IO Block: 4096   regular file
Device: 254,0   Inode: 35254       Links: 1
Access: (0555/-r-xr-xr-x)  Uid: (    0/    root)   Gid: (    0/ UNKNOWN)
Access: 2024-08-21 23:24:56.307000000 +0000
Modify: 2024-08-21 23:24:56.216000000 +0000
Change: 2024-08-21 23:24:56.216000000 +0000
 Birth: 2024-08-21 23:24:56.216000000 +0000
```

And the content has not been modified (the diff is empty).

**Comment by philiptaron** [comment]
If there are no blockers, I move that we merge this within the next day.

**Pull request merged**